pandas/core/ops/__init__.py

```diff
@@ -340,27 +336,24 @@ def dispatch_to_series(left, right, func, str_rep=None, axis=None):
             # Note: we do not do this unconditionally as it may be lossy or
             #  expensive for EA dtypes.
             right = np.asarray(right)
-
-            def column_op(a, b):
-                return {i: func(a.iloc[:, i], b[i]) for i in range(len(a.columns))}
-
         else:
+            right = right._values
 
-            def column_op(a, b):
-                return {i: func(a.iloc[:, i], b.iloc[i]) for i in range(len(a.columns))}
+        arrays = [array_op(l, r) for l, r in zip(left._iter_column_arrays(), right)]
 
     elif isinstance(right, ABCSeries):
         assert right.index.equals(left.index)  # Handle other cases later
+        right = right._values
 
-        def column_op(a, b):
-            return {i: func(a.iloc[:, i], b) for i in range(len(a.columns))}
+        arrays = [array_op(l, right) for l in left._iter_column_arrays()]
 
     else:
         # Remaining cases have less-obvious dispatch rules
         raise NotImplementedError(right)
 
-    new_data = expressions.evaluate(column_op, str_rep, left, right)
-    return new_data
+    return type(left)._from_arrays(
```

Comment:
i think you can just return `arrays` here, since that will be passed to `left._construct_result` (though the docstring does say DataFrame, i guess thats wrong ATM)

Reply:
I want to explicitly use `_from_arrays`, as that is part of the reason that makes this PR a perf improvement (the `_from_arrays` constructor specifically handles the case of a list of arrays corresponding to columns). Also, the default DataFrame constructor handles a list of arrays differently.

Reply:
sounds good